yt_dlp/networking/_helper.py

```diff
@@ -206,3 +207,54 @@ def wrapper(self, *args, **kwargs):
                 e.handler = self
             raise
     return wrapper
+
+
+def _socket_connect(ip_addr, timeout, source_address):
+    af, socktype, proto, canonname, sa = ip_addr
+    sock = socket.socket(af, socktype, proto)
+    try:
+        if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+            sock.settimeout(timeout)
+        if source_address:
+            sock.bind(source_address)
+        sock.connect(sa)
+        return sock
+    except socket.error:
+        sock.close()
+        raise
+
+
+def create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, source_address=None, *, _create_socket_func=_socket_connect):
+    # This is to workaround _create_connection() from socket where it will try all
+    # address data from getaddrinfo() including IPv6. This filters the result from
+    # getaddrinfo() based on the source_address value.
+    # This is based on the cpython socket.create_connection() function.
+    # https://github.com/python/cpython/blob/main/Lib/socket.py#L810
+    host, port = address
+    err = None
+    ip_addrs = addrs = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM)
```

Comment:
this is like this for the case getaddrinfo returns an empty list. we don't want to raise this error in that case

Reply:
Thats why the above change is needed. We `raise socket.error('getaddrinfo returns an empty list')` immadiately, so empty list will never reach this point. That allows us to just see if the filter is empty.

Reply:
What I mean is when source address if supplied and getaddrinfo returns an empty list, it should raise the getaddrinfo error.

Reply:
It should raise the error *always* when getaddrinfo returns an empty list. Please see again this path:
```py
    ip_addrs = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM)
    if not ip_addrs:
        raise socket.error('getaddrinfo returns an empty list')
```
We raise that error early, instead of going through the `if` branch, then through the `for` that doesn't loop and then checking `if err is None`. This is identical from behavior, since if source address is set or not, it will always raise that error anyways. Please trace through it in your head again to see that all places somehow fallthrough to the `raise`.

We can therefore lift the `raise` and since we dont have to fall through to that `getaddrinfo` error we can also simplify the source address branch.

Reply:
oh lmfao I am blind, I didn't see you moved the error up there